File: codescout-companion/scripts/detect.py
#!/usr/bin/env python3
"""Codescout detection logic, factored out of detect-tools.sh.

Reads env (CWD, HOME, CLAUDE_CONFIG_DIR) + filesystem, emits shell-eval lines
on stdout. Hooks consume via:

    eval "$(CWD=$CWD python3 .../scripts/detect.py)"

Variables emitted (matching the legacy detect-tools.sh contract byte-for-byte
where the characterization test asserts equality):

    HAS_CODESCOUT, CS_SERVER_NAME, CS_PREFIX, CS_BINARY, CS_PROJECT_DIR,
    HAS_CS_ONBOARDING, HAS_CS_MEMORIES, CS_MEMORY_NAMES,
    HAS_CS_SYSTEM_PROMPT, CS_SYSTEM_PROMPT, BLOCK_READS, WORKSPACE_ROOT,
    SOURCE_EXT_PATTERN

Use --json for a structured dump (no shell quoting; for tests).
"""

from __future__ import annotations

import json
import os
import re
import shlex
import sys
from pathlib import Path
from typing import Any

SOURCE_EXT_PATTERN = r"\.(kt|kts|java|ts|tsx|js|jsx|py|go|rs|cs|rb|scala|swift|cpp|c|h|hpp|sh|bash)$"
SERVER_NAME_RE = re.compile(r"codescout")


def _load_json(path: Path) -> dict[str, Any] | None:
    try:
        with path.open("r", encoding="utf-8") as f:
            data = json.load(f)
        return data if isinstance(data, dict) else None
    except (OSError, json.JSONDecodeError):
        return None


def _expand_home(value: str, home: str) -> str:
    if value.startswith("~/"):
        return f"{home}{value[1:]}"
    if value == "~":
        return home
    return value


def _find_routing_config(cwd: Path) -> Path | None:
    candidates = [
        cwd / ".claude" / "codescout-companion.json",
        cwd / ".claude" / "codescout-routing.json",
    ]
    for p in candidates:
        if p.is_file():
            return p
    return None


def _find_project_dir(cwd: Path) -> Path:
    return cwd / ".codescout"


def _server_name_from_mcp_config(cfg: dict[str, Any]) -> str:
    servers = cfg.get("mcpServers") or {}
    if not isinstance(servers, dict):
        return ""
    for key, value in servers.items():
        if not isinstance(value, dict):
            continue
        cmd = value.get("command", "") or ""
        if isinstance(cmd, str) and SERVER_NAME_RE.search(cmd):
            return key
        args = value.get("args") or []
        if isinstance(args, list) and any(
            isinstance(a, str) and SERVER_NAME_RE.search(a) for a in args
        ):
            return key
    return ""


def _extract_command(cfg: dict[str, Any], server_name: str) -> str:
    servers = cfg.get("mcpServers") or {}
    if not isinstance(servers, dict):
        return ""
    entry = servers.get(server_name)
    if not isinstance(entry, dict):
        return ""
    cmd = entry.get("command", "") or ""
    return cmd if isinstance(cmd, str) else ""


def detect(cwd: str, home: str, claude_config_dir: str | None) -> dict[str, str]:
    cwd_path = Path(cwd)

    routing_path = _find_routing_config(cwd_path)
    project_dir = _find_project_dir(cwd_path)
    memories_dir = project_dir / "memories"
    config_file = project_dir / "project.toml"
    system_prompt_file = project_dir / "system-prompt.md"
    mcp_json = cwd_path / ".mcp.json"

    has_codescout = False
    server_name = ""

    routing_cfg: dict[str, Any] = {}
    if routing_path is not None:
        routing_cfg = _load_json(routing_path) or {}
        override = routing_cfg.get("server_name")
        if isinstance(override, str) and override:
            has_codescout = True
            server_name = override

    if not has_codescout and mcp_json.is_file():
        cfg = _load_json(mcp_json)
        if cfg:
            name = _server_name_from_mcp_config(cfg)
            if name:
                has_codescout = True
                server_name = name

    claude_dir = Path(claude_config_dir) if claude_config_dir else Path(home) / ".claude"
    user_configs: list[Path] = [
        claude_dir / ".claude.json",
        claude_dir / "settings.json",
    ]
    if not claude_config_dir:
        user_configs.append(Path(home) / ".claude.json")

    if not has_codescout:
        for cfg_path in user_configs:
            if not cfg_path.is_file():
                continue
            cfg = _load_json(cfg_path)
            if not cfg:
                continue
            name = _server_name_from_mcp_config(cfg)
            if name:
                has_codescout = True
                server_name = name
                break

    prefix = f"mcp__{server_name}__" if has_codescout else ""

    binary = ""
    if has_codescout and server_name:
        binary_search = [mcp_json, claude_dir / ".claude.json", claude_dir / "settings.json", Path(home) / ".claude.json"]
        for cfg_path in binary_search:
            if not cfg_path.is_file():
                continue
            cfg = _load_json(cfg_path)
            if not cfg:
                continue
            cmd = _extract_command(cfg, server_name)
            if cmd:
                binary = _expand_home(cmd, home)
                break

    block_reads = "true"
    workspace_root = ""
    if routing_cfg:
        block_val = routing_cfg.get("block_reads")
        if block_val is False or (isinstance(block_val, str) and block_val == "false"):
            block_reads = "false"
        ws = routing_cfg.get("workspace_root")
        if isinstance(ws, str) and ws:
            workspace_root = _expand_home(ws, home)

    has_onboarding = "true" if config_file.is_file() else "false"

    has_memories = "false"
    memory_names = ""
    if memories_dir.is_dir():
        for entry in sorted(memories_dir.iterdir()):
            if entry.is_file() and entry.suffix == ".md":
                memory_names += f"{entry.stem} "
                has_memories = "true"

    has_system_prompt = "false"
    system_prompt = ""
    if system_prompt_file.is_file():
        try:
            system_prompt = system_prompt_file.read_text(encoding="utf-8")
            has_system_prompt = "true"
        except OSError:
            pass

    return {
        "HAS_CODESCOUT": "true" if has_codescout else "false",
        "CS_SERVER_NAME": server_name,
        "CS_PREFIX": prefix,
        "CS_BINARY": binary,
        "CS_PROJECT_DIR": str(project_dir),
        "CS_MEMORIES_DIR": str(memories_dir),
        "CS_CONFIG_FILE": str(config_file),
        "ROUTING_CONFIG": str(routing_path) if routing_path else "",
        "HAS_CS_ONBOARDING": has_onboarding,
        "HAS_CS_MEMORIES": has_memories,
        "CS_MEMORY_NAMES": memory_names,
        "HAS_CS_SYSTEM_PROMPT": has_system_prompt,
        "CS_SYSTEM_PROMPT": system_prompt,
        "BLOCK_READS": block_reads,
        "WORKSPACE_ROOT": workspace_root,
        "SOURCE_EXT_PATTERN": SOURCE_EXT_PATTERN,
    }


def emit_shell(values: dict[str, str]) -> None:
    for key, value in values.items():
        sys.stdout.write(f"{key}={shlex.quote(value)}\n")


def emit_json(values: dict[str, str]) -> None:
    json.dump(values, sys.stdout, indent=2)
    sys.stdout.write("\n")


def main(argv: list[str]) -> int:
    json_mode = "--json" in argv

    cwd = os.environ.get("CWD", "")
    home = os.environ.get("HOME", "")
    claude_config_dir = os.environ.get("CLAUDE_CONFIG_DIR") or None

    if not cwd:
        print("CWD must be set in env", file=sys.stderr)
        return 1

    values = detect(cwd, home, claude_config_dir)

    if json_mode:
        emit_json(values)
    else:
        emit_shell(values)
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: codescout-companion/hooks/detect.mjs
// codescout-companion/hooks/detect.mjs
// JS port of scripts/detect.py — the codescout detection foundation shared by the
// Node hooks. Node-only (no Python needed on Windows). Kept byte-parity with
// detect.py (compared via --json) by hooks/detect.test.sh.
//
// Parity caveat: Python json.dump defaults to ensure_ascii=True; JSON.stringify
// emits UTF-8. Parity holds for ASCII values (which the tests use); non-ASCII
// system-prompt/memory content would diverge in escaping only.
import { readFileSync, statSync, readdirSync } from 'node:fs';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';

export const SOURCE_EXT_PATTERN =
  '\\.(kt|kts|java|ts|tsx|js|jsx|py|go|rs|cs|rb|scala|swift|cpp|c|h|hpp|sh|bash)$';
const SERVER_NAME_RE = /codescout/;

function isFile(p) {
  try {
    return statSync(p).isFile();
  } catch {
    return false;
  }
}

function isDir(p) {
  try {
    return statSync(p).isDirectory();
  } catch {
    return false;
  }
}

function loadJson(path) {
  try {
    const data = JSON.parse(readFileSync(path, 'utf8'));
    return data && typeof data === 'object' && !Array.isArray(data) ? data : null;
  } catch {
    return null;
  }
}

function expandHome(value, home) {
  if (value.startsWith('~/')) return `${home}${value.slice(1)}`;
  if (value === '~') return home;
  return value;
}

function findRoutingConfig(cwd) {
  const candidates = [
    join(cwd, '.claude', 'codescout-companion.json'),
    join(cwd, '.claude', 'codescout-routing.json'),
  ];
  for (const p of candidates) if (isFile(p)) return p;
  return null;
}

function serverNameFromMcpConfig(cfg) {
  const servers = cfg.mcpServers;
  if (!servers || typeof servers !== 'object' || Array.isArray(servers)) return '';
  for (const [key, value] of Object.entries(servers)) {
    if (!value || typeof value !== 'object' || Array.isArray(value)) continue;
    const cmd = value.command || '';
    if (typeof cmd === 'string' && SERVER_NAME_RE.test(cmd)) return key;
    const args = value.args || [];
    if (Array.isArray(args) && args.some((a) => typeof a === 'string' && SERVER_NAME_RE.test(a))) {
      return key;
    }
  }
  return '';
}

function extractCommand(cfg, serverName) {
  const servers = cfg.mcpServers;
  if (!servers || typeof servers !== 'object' || Array.isArray(servers)) return '';
  const entry = servers[serverName];
  if (!entry || typeof entry !== 'object' || Array.isArray(entry)) return '';
  const cmd = entry.command || '';
  return typeof cmd === 'string' ? cmd : '';
}

export function detect(cwd, home, claudeConfigDir) {
  const projectDir = join(cwd, '.codescout');
  const memoriesDir = join(projectDir, 'memories');
  const configFile = join(projectDir, 'project.toml');
  const systemPromptFile = join(projectDir, 'system-prompt.md');
  const mcpJson = join(cwd, '.mcp.json');

  const routingPath = findRoutingConfig(cwd);
  let hasCodescout = false;
  let serverName = '';

  let routingCfg = {};
  if (routingPath) {
    routingCfg = loadJson(routingPath) || {};
    const override = routingCfg.server_name;
    if (typeof override === 'string' && override) {
      hasCodescout = true;
      serverName = override;
    }
  }

  if (!hasCodescout && isFile(mcpJson)) {
    const cfg = loadJson(mcpJson);
    if (cfg) {
      const name = serverNameFromMcpConfig(cfg);
      if (name) {
        hasCodescout = true;
        serverName = name;
      }
    }
  }

  const claudeDir = claudeConfigDir ? claudeConfigDir : join(home, '.claude');
  const userConfigs = [join(claudeDir, '.claude.json'), join(claudeDir, 'settings.json')];
  if (!claudeConfigDir) userConfigs.push(join(home, '.claude.json'));

  if (!hasCodescout) {
    for (const cfgPath of userConfigs) {
      if (!isFile(cfgPath)) continue;
      const cfg = loadJson(cfgPath);
      if (!cfg) continue;
      const name = serverNameFromMcpConfig(cfg);
      if (name) {
        hasCodescout = true;
        serverName = name;
        break;
      }
    }
  }

  const prefix = hasCodescout ? `mcp__${serverName}__` : '';

  let binary = '';
  if (hasCodescout && serverName) {
    const binarySearch = [
      mcpJson,
      join(claudeDir, '.claude.json'),
      join(claudeDir, 'settings.json'),
      join(home, '.claude.json'),
    ];
    for (const cfgPath of binarySearch) {
      if (!isFile(cfgPath)) continue;
      const cfg = loadJson(cfgPath);
      if (!cfg) continue;
      const cmd = extractCommand(cfg, serverName);
      if (cmd) {
        binary = expandHome(cmd, home);
        break;
      }
    }
  }

  let blockReads = 'true';
  let workspaceRoot = '';
  if (routingCfg && Object.keys(routingCfg).length) {
    const blockVal = routingCfg.block_reads;
    if (blockVal === false || blockVal === 'false') blockReads = 'false';
    const ws = routingCfg.workspace_root;
    if (typeof ws === 'string' && ws) workspaceRoot = expandHome(ws, home);
  }

  const hasOnboarding = isFile(configFile) ? 'true' : 'false';

  let hasMemories = 'false';
  let memoryNames = '';
  if (isDir(memoriesDir)) {
    // Recurses, because a memory topic is a PATH. The server's authority is
    // MemoryStore::list (codescout src/memory/mod.rs:114): it walks with walkdir at
    // arbitrary depth and names each topic by its path relative to the memories dir,
    // extension stripped, separators normalised to '/'. That string is the topic key —
    // memory(action="read", topic="infra/friction-measurement") — so a shallower scan
    // here does not merely under-report, it makes those topics unreachable through
    // every surface fed by CS_MEMORY_NAMES: the SessionStart banner, and the subagent
    // Phase 0 block, which substitutes this list FOR the memory(action="list") call.
    // docs/issues/archive/2026-08-27-cs-memory-names-skips-namespaced-memories.md
    //
    // withFileTypes, not statSync: Dirent's isFile()/isDirectory() are lstat-based and
    // BOTH false for a symlink, which reproduces walkdir's no-follow default and makes
    // a symlink cycle under memories/ unable to hang this hook. statSync follows links,
    // so the obvious recursion written with the isFile/isDir helpers above can loop —
    // and this runs at SessionStart, where a hang costs the whole session.
    const walk = (dir, prefix) => {
      let entries;
      try {
        entries = readdirSync(dir, { withFileTypes: true });
      } catch {
        // fail-open PER DIRECTORY: one unreadable subdirectory hides itself, never its
        // siblings. The pre-recursion version wrapped the whole scan, so a single
        // failure reported zero memories.
        return;
      }
      entries.sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
      for (const e of entries) {
        const name = e.name;
        // name.length > 3 excludes a file literally named ".md" (parity with
        // detect.py, where Path(".md").suffix == "" so it is not a memory).
        if (e.isFile() && name.endsWith('.md') && name.length > 3) {
          memoryNames += `${prefix}${name.slice(0, -3)} `;
          hasMemories = 'true';
        } else if (e.isDirectory()) {
          walk(join(dir, name), `${prefix}${name}/`);
        }
      }
    };
    walk(memoriesDir, '');
  }

  let hasSystemPrompt = 'false';
  let systemPrompt = '';
  if (isFile(systemPromptFile)) {
    try {
      systemPrompt = readFileSync(systemPromptFile, 'utf8');
      hasSystemPrompt = 'true';
    } catch {
      /* fail-open */
    }
  }

  return {
    HAS_CODESCOUT: hasCodescout ? 'true' : 'false',
    CS_SERVER_NAME: serverName,
    CS_PREFIX: prefix,
    CS_BINARY: binary,
    CS_PROJECT_DIR: projectDir,
    CS_MEMORIES_DIR: memoriesDir,
    CS_CONFIG_FILE: configFile,
    ROUTING_CONFIG: routingPath || '',
    HAS_CS_ONBOARDING: hasOnboarding,
    HAS_CS_MEMORIES: hasMemories,
    CS_MEMORY_NAMES: memoryNames,
    HAS_CS_SYSTEM_PROMPT: hasSystemPrompt,
    CS_SYSTEM_PROMPT: systemPrompt,
    BLOCK_READS: blockReads,
    WORKSPACE_ROOT: workspaceRoot,
    SOURCE_EXT_PATTERN: SOURCE_EXT_PATTERN,
  };
}

// CLI (parity testing only): `CWD=... node detect.mjs --json`.
if (process.argv[1] && fileURLToPath(import.meta.url) === process.argv[1]) {
  const cwd = process.env.CWD || '';
  const home = process.env.HOME || '';
  const claudeConfigDir = process.env.CLAUDE_CONFIG_DIR || null;
  if (!cwd) {
    process.stderr.write('CWD must be set in env\n');
    process.exit(1);
  }
  process.stdout.write(JSON.stringify(detect(cwd, home, claudeConfigDir), null, 2) + '\n');
}
